flashlight/autograd/backend/cuda/BatchNorm.cpp

```diff
@@ -131,70 +136,80 @@ Variable batchnorm(
-        {
-          DevicePtr i_raw(in.array());
-          DevicePtr w_raw(wt.array());
-
-          DevicePtr grad_in_raw(grad_in.array());
-          DevicePtr grad_wt_raw(grad_wt.array());
-          DevicePtr grad_bs_raw(grad_bs.array());
-
-          DevicePtr grad_op_raw(grad_output.array());
-
-          DevicePtr save_mean_raw(save_mean);
-          DevicePtr save_var_raw(save_var);
-
-          CUDNN_CHECK_ERR(cudnnBatchNormalizationBackward(
-              getCudnnHandle(),
-              mode,
-              one1,
-              zero0,
-              one1,
-              zero0,
-              i_desc.descriptor,
-              i_raw.get(),
-              i_desc.descriptor,
-              grad_op_raw.get(),
-              i_desc.descriptor,
-              grad_in_raw.get(),
-              w_desc.descriptor,
-              w_raw.get(),
-              grad_wt_raw.get(),
-              grad_bs_raw.get(),
-              epsilon,
-              save_mean_raw.get(),
-              save_var_raw.get()));
-        }
-        in.addGrad(grad_in);
-        wt.addGrad(grad_wt);
-        if (!bs.isempty()) {
-          bs.addGrad(grad_bs);
-        }
-      };
+  auto gradFunc = [train,
+                   save_mean,
+                   save_var,
+                   mode,
+                   in_desc_dims,
+                   wt_desc_dims,
+                   epsilon,
+                   useMixedPrecision](
+                      std::vector<Variable>& inputs,
+                      const Variable& grad_output) {
+    typeTrace("Batchnorm BWD", grad_output.type());
+    grad_output.autoCast(useMixedPrecision);
+    if (!train) {
+      throw std::logic_error(
+          "can't compute batchnorm grad when train was not specified");
+    }
+
+    auto& in = inputs[0];
+    auto wt = inputs[1].isempty()
+        ? Variable(af::constant(1.0, wt_desc_dims, in.type()), false)
+        : inputs[1];
+    auto& bs = inputs[2];
+
+    auto scalarsType = in.type() == f16 ? f32 : in.type();
+    const void* one1 = kOne(scalarsType);
+    const void* zero0 = kZero(scalarsType);
+
+    auto i_desc = TensorDescriptor(in.type(), in_desc_dims);
+    auto w_desc = TensorDescriptor(wt.type(), wt_desc_dims);
+    // CuDNN doesn't support calculating only the gradients
+    // required for batchnorm
+    auto grad_in = Variable(af::array(in.dims(), in.type()), false);
+    auto grad_wt = Variable(af::array(wt.dims(), wt.type()), false);
+    auto grad_bs = Variable(af::array(wt.dims(), wt.type()), false);
+    {
+      DevicePtr i_raw(in.array());
+      DevicePtr w_raw(wt.array());
+
+      DevicePtr grad_in_raw(grad_in.array());
+      DevicePtr grad_wt_raw(grad_wt.array());
+      DevicePtr grad_bs_raw(grad_bs.array());
+
+      DevicePtr grad_op_raw(grad_output.array());
+
+      DevicePtr save_mean_raw(save_mean);
+      DevicePtr save_var_raw(save_var);
+
+      CUDNN_CHECK_ERR(cudnnBatchNormalizationBackward(
```

Comment:
Noticing that this logic hasn't changed at all for the AMP case - is this intentional?

Reply:
Actually for BN, if I understand your question correctly, providing input tensors with correct type is all we need to do. In the current version, we take care of that with the `autoCast`.

Reply:
Ok. Once we remove `autoCast`, the tensor `DevicePtr` should already have the correct precision.